README.md
# shadowtunnel

## Introduce

shadowtunnel is a secure encryption tunnel between your local machine and remote service to protect your TCP flow，

which can efficiently compress transmission, and the flow has no characteristics.

local machine <----> shadowtunnel <----> service on remote.

## Usage

```text
Usage of ./shadowtunnel:
  -E    outbound connection is encrypted
  -U    outbound connection is udp
  -c    compress traffic
  -cache string
        dns query cache file path (default "cache.dat")
  -daemon
        daemon mode
  -debug
        show debug info
  -dns string
        local dns server listen on address
  -dns-hosts string
        path of dns hosts file
  -dns-proxy
        is dns endpoint or not
  -dns-server string
        remote dns server to resolve domain (default "8.8.8.8:53")
  -e    inbound connection is encrypted
  -f weight
        forward address,such as : 127.0.0.1:8080 or with @weight: 127.0.0.1:8080@1
  -forever
        forever mode
  -l string
        local listen address, such as : 0.0.0.0:33000 (default ":50000")
  -lb-activeafter int
        host going actived after this success count (default 1)
  -lb-hashtarget hash
        use target address to choose parent for LB, only worked for LB's hash
        method and using `-redir` (default true)
  -lb-inactiveafter int
        host going inactived after this fail count (default 2)
  -lb-method string
        load balance method when use multiple parent,can be
        <roundrobin|leastconn|leasttime|hash|weight> (default "leasttime")
  -lb-onlyha high availability mode
        use only high availability mode to choose parent for LB
  -lb-retrytime int
        sleep time milliseconds after checking (default 2000)
  -lb-timeout int
        tcp milliseconds timeout of connecting to parent (default 3000)
  -log string
        logging output to file
  -m string
        method of encrypt/decrypt, these below are supported :
        aes-128-cfb,aes-192-cfb,des-cfb,cast5-cfb,rc4-md5,chacha20,aes-256-cfb,
        aes-128-ctr,aes-192-ctr,aes-256-ctr,bf-cfb,rc4-md5-6,chacha20-ietf
        (default "aes-192-cfb")
  -nolog
        turn off logging
  -p string
        password of encrypt/decrypt (default "shadowtunnel")
  -profiling
        profiling mode, in this mode, you should stopping process
        by : Ctrl+C or 'kill -s SIGHUP $PID_OF_shadowtunnel'
  -redir
        read target from socket's redirect opts of iptables
  -t int
        connection timeout seconds (default 3)
  -ttl int
        cache seconds of dns query , if zero , default ttl used. (default 300)
  -u    inbound connection is udp
  -v    show version
```

## for example

1.http proxy

if we have a vps, IP is 2.2.2.2

firstly, we start a http proxy on 2.2.2.2.    
download http proxy program, execute below on line command on vps with root：

wget https://github.com/snail007/goproxy/releases/download/v4.9/proxy-linux-amd64.tar.gz && tar zxfv proxy-linux-amd64.tar.gz && rm proxy-linux-amd64.tar.gz && mv proxy /usr/bin/ && proxy http -p 127.0.0.1:38080 &

2.Start a tunnel on VPS

download shadowtunnel program, execute below on line command on vps with root：

wget https://github.com/snail007/shadowtunnel/releases/download/v1.1/shadowtunnel-linux-amd64.tar.gz && tar zxfv shadowtunnel-linux-amd64.tar.gz && rm shadowtunnel-linux-amd64.tar.gz && mv shadowtunnel /usr/bin/

start a tunnel on vps listening on :50000 and forward to 127.0.0.1:38080 ：

`shadowtunnel -e -f 127.0.0.1:38080 -l :50000`

3.start a tunnel on local machine

start a tunnel on local machine listening on :50000 and forward to 2.2.2.2:50000 :

`shadowtunnel -E -f 2.2.2.2:50000 -l :50000`

4.set http proxy configuration in chrome

setting local chrome's http proxy configuration as below：

ip: 127.0.0.1
port: 50000

5.done

## TCP over UDP

1.http proxy

if we have a vps, IP is 2.2.2.2.   

firstly, we start a http proxy on 2.2.2.2.    
download http proxy program, execute below on line command on vps with root：

`wget https://github.com/snail007/goproxy/releases/download/v4.9/proxy-linux-amd64.tar.gz && tar zxfv proxy-linux-amd64.tar.gz && rm proxy-linux-amd64.tar.gz && mv proxy /usr/bin/ && proxy http -p 127.0.0.1:38080 &`

2.start a tunnel on vps

download shadowtunnel program, execute below on line command on vps with root：

`wget https://github.com/snail007/shadowtunnel/releases/download/v1.1/shadowtunnel-linux-amd64.tar.gz && tar zxfv shadowtunnel-linux-amd64.tar.gz && rm shadowtunnel-linux-amd64.tar.gz && mv shadowtunnel /usr/bin/`

start a tunnel on vps listening on udp :50000 and forward to tcp 127.0.0.1:38080 ：

`shadowtunnel -u -e -f 127.0.0.1:38080 -l :50000 -p your-password`

3.start a tunnel on local machine

start a tunnel on local machine listening on tcp :50000 and forward to udp 2.2.2.2:50000 :

`shadowtunnel -U -E -f 2.2.2.2:50000 -l :50000 -p your-password`

4.set http proxy configuration in chrome

setting local chrome's http proxy configuration as below：

ip: 127.0.0.1
port: 50000

5.Done

## Deamon & Forever & Log

-daemon:

Using the parameter -daemon allows shadowtunnel to detached from the current command line and run in the background.

-forever:

Using parameter -forever allows shadowtunnel to run in the way of creating and monitoring child processes,

If an abnormal exit occurs, the child process will be restarted to ensure that the service is always online..

-log

Using parameter -log, you can set the log output to the file instead of the command line output.

-nolog

Using parameter -nolog can completely shut off log output and save CPU occupation.

Generally, the three parameters (-daemon -forever -log /tmp/st.log) are used together, so that we can find out the cause of the problem by looking at the log when we have a problem.

for example:

`shadowtunnel -u -e -f 127.0.0.1:38080 -l :50000 -p your-password -daemon -forever -log /tmp/st.log`

## DNS SERVICE

Shadowtunnel can provide local DNS query service, and has caching function to improve resolution speed.

It is necessary to have a superior service to start a DNS service while launching the port forwarding locally.

The -dns parameter sets the IP and port which the local DNS service listen, for example:0.0.0.0:5353

The -dns-server Parameters can be set to DNS servers that are ultimately used to resolve domain names, requiring the server to support TCP-style DNS queries, default:8.8.8.8:53.

for example:

`shadowtunnel -E -f 2.2.2.2:50000 -l :50000  -p your-password -dns :5353 -dns-server 8.8.8.8:53`

superior example:

If the superior is chain-style, then the superior of the DNS proxy in the chain is required to add the -dns-proxy parameter.

`shadowtunnel -e -f 127.0.0.1:38080 -l :50000 -p your-password -dns-proxy`

## DNS CACHE

The -ttl parameter can set the DNS query result cache time. unit is second. if it is 0, and use the TTL of the query result.

The -cache parameter sets DNS cache file location to prevent program restart and cache disappear, which will reduce performance.    

## LOAD BALANCE

Support superior load balancing, repeat -f parameters if exist multiple superiors.

`shadowtunnel -E -f 2.2.2.2:50000 -f 3.3.3.3:50000 -l :50000`

### SET RETRY INTERVAL AND TIMEOUT TIME

`shadowtunnel -E -f 2.2.2.2:50000 -f 3.3.3.3:50000 -l :50000 -lb-method leastconn -lb-retrytime 300 -lb-timeout 300`

### SETTING WEIGHT  

`shadowtunnel -E -f 2.2.2.2:50000@2 -f 3.3.3.3:50000@1 -l :50000 -lb-method weight -lb-retrytime 300 -lb-timeout 300`

### USE TARGET ADDRESS TO SELECT SUPERIOR

`shadowtunnel -E -f 2.2.2.2:50000@2 -f 3.3.3.3:50000@1 -l :50000 -lb-method hash -lb-hashtarget -lb-retrytime 300 -lb-timeout 300`

## DNS HOSTS

The -dns- hosts parameter sets the hosts file to be used when DNS is parsed, with the same content format as the system's hosts file.

`shadowtunnel -f 2.2.2.2:50000 -dns :5353 -dns-hosts /etc/hosts`


CHANGELOG
shadowtunnel更新日志

v1.7
1.权重标志@更改为#
2.上级支持单独设置密码
3.重构了代码,提升了效率.
4.修复了无法stop的bug.

v1.6
1.增加了sdk.

v1.5
1.优化了dns转发规则的格式,增加了相同的dns可省略的支持,如果没有指定dns地址,使用最后一次有效的dns地址.
比如:
a.com 114.114.114.114
b.com
c.com
d.com 8.8.8.8
e.com
那么:
b.com,c.com使用的是:114.114.114.114
e.com使用的是:8.8.8.8

v1.4
1.优化了dns的域名匹配,增加了匹配缓存,匹配更快速.

v1.3
1.新增-dns-hosts参数,对hosts文件支持.
2.新增-dns-forward域名列表转发解析支持,可以指定通配符域名使用指定的DNS服务器解析.


v1.2
1.新增了-daemon参数,后台运行功能.
2.新增了-forever参数,永久运行功能,异常退出会自己重启.
3.新增了-log参数,设置日志输出到文件.
4.新增了-profiling参数,调试分析功能,用于遇到性能问题的时候可以获取调试信息用于解决问题.
5.新增了-nolog参数,禁用日志输出功能,可以节省CPU.
6.新增了DNS代理功能,提供安全的DNS解析,详细信息参考手册DNS部分.
7.新增了负载均衡高可用功能,详细情况查看命令帮助-lb开头参数.


v1.1
1.新增了TCP over UDP功能.


core/st.go
package core

import (
	"bufio"
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"flag"
	"fmt"
	"io/ioutil"
	logger "log"
	"net"
	"os"
	"os/exec"
	"os/signal"
	"path"
	"path/filepath"
	"runtime/debug"
	"runtime/pprof"
	"strconv"
	"strings"
	"syscall"
	"time"

	"sync"

	"github.com/gobwas/glob"
	"github.com/miekg/dns"
	gocache "github.com/pmylund/go-cache"
	clienttransport "github.com/snail007/proxy/core/cs/client"
	srvtransport "github.com/snail007/proxy/core/cs/server"
	tou "github.com/snail007/proxy/core/dst"
	encryptconn "github.com/snail007/proxy/core/lib/transport/encrypt"
	utils "github.com/snail007/proxy/utils"
	jumper "github.com/snail007/proxy/utils/jumper"
	lbx "github.com/snail007/proxy/utils/lb"
	redirx "github.com/snail007/shadowtunnel/redir"
)

const (
	VERSION = "1.7"
)

type forwarders []string

func (i *forwarders) String() string {
	return strings.Join(*i, ",")
}

func (i *forwarders) Set(value string) error {
	*i = append(*i, value)
	return nil
}

var (
	timer             *time.Ticker
	listenAddr        string
	forwardsAddr      forwarders
	timeout           int
	compress          bool
	method            string
	password          string
	listen            srvtransport.ServerChannel
	err               error
	inboundEncrypt    bool
	outboundEncrypt   bool
	inboundUDP        bool
	outboundUDP       bool
	version           bool
	dnsListen         string
	dnsServerAddress  string
	dnsProxy          bool
	dnsListenerServer *dns.Server
	dnsTTL            int
	cache             *gocache.Cache
	cacheFile         string
	dialer            jumper.Jumper

	hosts          string
	dnsHosts       = map[string]string{}
	dnsForwardFile string
	dnsForward     = map[string]string{}
	redir          bool

	//LoadBalance
	lb                       lbx.Group
	loadBalanceMethod        string
	loadBalanceTimeout       int
	loadBalanceRetryTime     int
	loadBalanceHashTarget    bool
	loadBalanceOnlyHA        bool
	loadBalanceActiveAfter   int
	loadBalanceInactiveAfter int

	//common
	isDebug   bool
	nolog     bool
	log       = logger.New(os.Stderr, "", logger.Ldate|logger.Ltime)
	logfile   string
	daemon    bool
	forever   bool
	profiling bool
	cmd       *exec.Cmd
	cpuProfilingFile,
	memProfilingFile,
	blockProfilingFile,
	goroutineProfilingFile,
	threadcreateProfilingFile *os.File

	dnsGlobCache = sync.Map{}

	conns = sync.Map{}
)

func Start(args string) {
	if args != "" {
		_args := strings.Fields(args)
		_fullArgs := []string{os.Args[0]}
		os.Args = append(_fullArgs, _args...)
	}
	flag.StringVar(&listenAddr, "l", ":50000", "local listen address, such as : 0.0.0.0:33000")
	flag.StringVar(&method, "m", "aes-192-cfb", "method of encrypt/decrypt, these below are supported :\n"+strings.Join(encryptconn.GetCipherMethods(), ","))
	flag.StringVar(&password, "p", "shadowtunnel", "password of encrypt/decrypt")
	flag.Var(&forwardsAddr, "f", "forward address,such as : 127.0.0.1:8080 or with #`weight`: 127.0.0.1:8080#1 or with `password`: password@127.0.0.1:8080#1")
	flag.IntVar(&timeout, "t", 3, "connection timeout seconds")
	flag.BoolVar(&compress, "c", false, "compress traffic")
	flag.BoolVar(&inboundEncrypt, "e", false, "inbound connection is encrypted")
	flag.BoolVar(&outboundEncrypt, "E", false, "outbound connection is encrypted")
	flag.BoolVar(&inboundUDP, "u", false, "inbound connection is udp")
	flag.BoolVar(&outboundUDP, "U", false, "outbound connection is udp")
	flag.StringVar(&hosts, "dns-hosts", "", "path of dns hosts file")
	flag.StringVar(&dnsForwardFile, "dns-forward", "", "rule file of resolving domain")
	flag.BoolVar(&redir, "redir", false, "read target from socket's redirect opts of iptables")
	//local
	flag.StringVar(&dnsListen, "dns", "", "local dns server listen on address")
	flag.StringVar(&dnsServerAddress, "dns-server", "8.8.8.8:53", "remote dns server to resolve domain")
	//server
	flag.BoolVar(&dnsProxy, "dns-proxy", false, "is dns endpoint or not")

	flag.IntVar(&dnsTTL, "ttl", 300, "cache seconds of dns query , if zero , default ttl used.")
	flag.StringVar(&cacheFile, "cache", filepath.Join(path.Dir(os.Args[0]), "cache.dat"), "dns query cache file path")
	flag.BoolVar(&version, "v", false, "show version")
	//lb
	flag.StringVar(&loadBalanceMethod, "lb-method", "leasttime", "load balance method when use multiple parent,can be <roundrobin|leastconn|leasttime|hash|weight>")
	flag.IntVar(&loadBalanceRetryTime, "lb-retrytime", 2000, "sleep time milliseconds after checking")
	flag.IntVar(&loadBalanceTimeout, "lb-timeout", 3000, "tcp milliseconds timeout of connecting to parent")
	flag.BoolVar(&loadBalanceHashTarget, "lb-hashtarget", true, "use target address to choose parent for LB, only worked for LB's `hash` method and using `-redir`")
	flag.BoolVar(&loadBalanceOnlyHA, "lb-onlyha", false, "use only `high availability mode` to choose parent for LB")
	flag.IntVar(&loadBalanceActiveAfter, "lb-activeafter", 1, "host going actived after this success count")
	flag.IntVar(&loadBalanceInactiveAfter, "lb-inactiveafter", 2, "host going inactived after this fail count")

	//common
	flag.BoolVar(&nolog, "nolog", false, "turn off logging")
	flag.BoolVar(&isDebug, "debug", false, "show debug info")
	flag.BoolVar(&daemon, "daemon", false, "daemon mode")
	flag.BoolVar(&forever, "forever", false, "forever mode")
	flag.BoolVar(&profiling, "profiling", false, "profiling mode, in this mode, you should stopping process by : Ctrl+C or 'kill -s SIGHUP $PID_OF_shadowtunnel'")
	flag.StringVar(&logfile, "log", "", "logging output to file")
	flag.Parse()
	if version {
		fmt.Println(VERSION)
		return
	}
	if len(forwardsAddr) == 0 || listenAddr == "" {
		flag.Usage()
		return
	}
	if args == "" && daemon {
		daemonF()
		return
	}
	if args == "" && forever {
		foreverF()
		return
	}
	if nolog {
		log.SetOutput(ioutil.Discard)
	}
	if profiling {
		startProfiling()
	}
	if outboundUDP && !outboundEncrypt {
		log.Fatal("outbound connection is udp , -E is required")
		return
	}
	if inboundUDP && !inboundEncrypt {
		log.Fatal("inbound connection is udp , -e is required")
		return
	}
	//setting lb
	initLB()
	//setting log
	tou.SetLogger(log)
	flags := logger.Ldate
	if isDebug {
		flags |= logger.Lshortfile | logger.Lmicroseconds
		log.SetFlags(flags)
	} else {
		flags |= logger.Ltime
	}
	if dnsListen != "" {
		//setting cache
		cache = gocache.New(time.Second*time.Duration(dnsTTL), time.Second*60)
		cache.LoadFile(cacheFile)
		go func() {
			timer = time.NewTicker(time.Second * 300)
			for {
				<-timer.C
				cache.DeleteExpired()
				cache.SaveFile(cacheFile)
			}
		}()
		//start dns
		initDnsHosts(hosts, &dnsHosts, ".")
		initDnsHosts(dnsForwardFile, &dnsForward, "")
		dnsServer()
	}
	listen = srvtransport.NewServerChannelHost(listenAddr, log)
	if inboundUDP {
		err = listen.ListenTOU(method, password, compress, callback)
	} else {
		if inboundEncrypt {
			err = listen.ListenTCPS(method, password, compress, callback)
		} else {
			err = listen.ListenTCP(callback)
		}
	}
	if err != nil {
		log.Fatal(err)
	}
	typ := "tcp"
	if inboundUDP {
		typ = "udp"
	}
	log.Printf("shadowtunnel listen on %s : %s", typ, listen.Addr())
	if args == "" {
		cleanup()
	}
}

func callback(conn net.Conn) {
	defer func() {
		if e := recover(); e != nil {
			debugf("callback handler crashed, err : %s\nstack:%s", e, string(debug.Stack()))
		}
	}()
	remoteAddr := conn.RemoteAddr()
	var outconn net.Conn
	var target string
	if dnsProxy {
		utils.ReadPacketData(conn, &target)
		if target == "" {
			//debugf("[warn] target is empty")
			conn.Close()
			return
		}
	} else {
		target = lb.Select("", loadBalanceOnlyHA)
	}
	if dnsProxy && target != "_" {
		outconn, err = net.DialTimeout("tcp", target, time.Duration(timeout)*time.Second)
	} else {
		if target == "_" {
			target = lb.Select("", loadBalanceOnlyHA)
		}
		addr := ""
		realAddress := ""
		if dnsListen != "" {
			addr = "_"
		}
		if redir {
			realAddress, err = redirx.RealServerAddress(&conn)
			if err != nil {
				debugf("%s <--> %s, error: %s", remoteAddr, target, err)
				conn.Close()
				return
			}
		}
		outconn, err = getOutconn(lb.Select(realAddress, loadBalanceOnlyHA), addr)
		if err == nil && redir {
			//debugf("real address %s", realAddress)
			pb := new(bytes.Buffer)
			pb.Write([]byte(fmt.Sprintf("CONNECT %s HTTP/1.1\r\n", realAddress)))
			pb.WriteString(fmt.Sprintf("Host: %s\r\n", realAddress))
			pb.WriteString(fmt.Sprintf("Proxy-Host: %s\r\n", realAddress))
			pb.WriteString("Proxy-Connection: Keep-Alive\r\n")
			pb.WriteString("Connection: Keep-Alive\r\n")
			pb.Write([]byte("\r\n"))
			_, err = outconn.Write(pb.Bytes())
			pb.Reset()
			pb = nil
			if err != nil {
				outconn.Close()
				conn.Close()
				conn = nil
				outconn = nil
				err = fmt.Errorf("error connecting to proxy: %s", err)
				return
			}
			reply := make([]byte, 1024)
			outconn.SetReadDeadline(time.Now().Add(time.Second * 5))
			n, err := outconn.Read(reply)
			outconn.SetReadDeadline(time.Time{})
			if err != nil {
				err = fmt.Errorf("error read reply from proxy: %s", err)
				outconn.Close()
				conn.Close()
				conn = nil
				outconn = nil
				return
			}
			if bytes.Index(reply[:n], []byte("200")) == -1 {
				err = fmt.Errorf("error greeting to proxy, response: %s", string(reply[:n]))
				outconn.Close()
				conn.Close()
				conn = nil
				outconn = nil
				return
			}
		}
	}
	if err != nil {
		debugf("%s <--> %s, error: %s", remoteAddr, target, err)
		conn.Close()
		return
	}
	key1 := outconn.LocalAddr().String()
	utils.IoBind(conn, outconn, func(err interface{}) {
		conns.Delete(remoteAddr)
		conns.Delete(key1)
		log.Printf("%s <--> %s released", remoteAddr, target)
	}, log)
	conns.Store(remoteAddr, &conn)
	conns.Store(key1, &outconn)
	log.Printf("%s <--> %s connected", remoteAddr, target)
}
func daemonF() {
	if daemon {
		args := []string{}
		for _, arg := range os.Args[1:] {
			if arg != "-daemon" {
				args = append(args, arg)
			}
		}
		cmd = exec.Command(os.Args[0], args...)
		cmd.Start()
		f := ""
		if forever {
			f = "forever "
		}
		debugf("%s%s [PID] %d running...\n", f, os.Args[0], cmd.Process.Pid)
		os.Exit(0)
	}
}
func foreverF() {
	args := []string{}
	for _, arg := range os.Args[1:] {
		if arg != "-forever" {
			args = append(args, arg)
		}
	}
	for {
		if cmd != nil {
			cmd.Process.Kill()
			time.Sleep(time.Second * 5)
		}
		cmd = exec.Command(os.Args[0], args...)
		cmdReaderStderr, err := cmd.StderrPipe()
		if err != nil {
			debugf("ERR:%s,restarting...\n", err)
			continue
		}
		cmdReader, err := cmd.StdoutPipe()
		if err != nil {
			debugf("ERR:%s,restarting...\n", err)
			continue
		}
		scanner := bufio.NewScanner(cmdReader)
		scannerStdErr := bufio.NewScanner(cmdReaderStderr)
		go func() {
			for scanner.Scan() {
				fmt.Println(scanner.Text())
			}
		}()
		go func() {
			for scannerStdErr.Scan() {
				fmt.Println(scannerStdErr.Text())
			}
		}()
		if err := cmd.Start(); err != nil {
			debugf("ERR:%s,restarting...\n", err)
			continue
		}
		pid := cmd.Process.Pid
		debugf("worker %s [PID] %d running...\n", os.Args[0], pid)
		if err := cmd.Wait(); err != nil {
			debugf("ERR:%s,restarting...", err)
			continue
		}
		debugf("worker %s [PID] %d unexpected exited, restarting...\n", os.Args[0], pid)
	}
}
func startProfiling() {
	cpuProfilingFile, _ = os.Create("cpu.prof")
	memProfilingFile, _ = os.Create("memory.prof")
	blockProfilingFile, _ = os.Create("block.prof")
	goroutineProfilingFile, _ = os.Create("goroutine.prof")
	threadcreateProfilingFile, _ = os.Create("threadcreate.prof")
	pprof.StartCPUProfile(cpuProfilingFile)
}
func stopProfiling() {
	goroutine := pprof.Lookup("goroutine")
	goroutine.WriteTo(goroutineProfilingFile, 1)
	heap := pprof.Lookup("heap")
	heap.WriteTo(memProfilingFile, 1)
	block := pprof.Lookup("block")
	block.WriteTo(blockProfilingFile, 1)
	threadcreate := pprof.Lookup("threadcreate")
	threadcreate.WriteTo(threadcreateProfilingFile, 1)
	pprof.StopCPUProfile()
}
func cleanup() {
	signalChan := make(chan os.Signal, 1)
	cleanupDone := make(chan bool)
	signal.Notify(signalChan,
		os.Interrupt,
		syscall.SIGHUP,
		syscall.SIGINT,
		syscall.SIGTERM,
		syscall.SIGQUIT)
	go func() {
		defer func() {
			if e := recover(); e != nil {
				debugf("clean handler crashed, err : \n%s", string(debug.Stack()))
			}
		}()
		for range signalChan {
			Stop()
			cleanupDone <- true
			break
		}
	}()
	<-cleanupDone
}

var sotpLock = &sync.Mutex{}
var stoped bool

func Stop() {
	sotpLock.Lock()
	defer sotpLock.Unlock()
	if !stoped {
		stoped = true
		listenAddr = ""
		if timer != nil {
			timer.Stop()
			timer = nil
		}
		if dnsListenerServer != nil {
			dnsListenerServer.Shutdown()
			dnsListenerServer = nil
		}
		lb.Stop()
		listen.Close()
		if profiling {
			stopProfiling()
		}
		keys := []interface{}{}
		conns.Range(func(k, v interface{}) bool {
			(*v.(*net.Conn)).Close()
			keys = append(keys, k)
			return true
		})
		for _, k := range keys {
			conns.Delete(k)
		}
	}
}
func initDnsHosts(file string, dnsHosts *map[string]string, subfix string) {
	hosts := file
	if hosts == "" {
		return
	}
	if utils.PathExists(hosts) {
		_content, err := ioutil.ReadFile(hosts)
		if err != nil {
			return
		}
		dnsHostArr := strings.Split(strings.Replace(string(_content), "\r", "", -1), "\n")
		n := 0
		last := ""
		for _, dnsHost := range dnsHostArr {
			dnsHost = strings.Trim(dnsHost, " \t")
			if strings.HasPrefix(dnsHost, "#") {
				continue
			}
			u := strings.Fields(strings.Trim(dnsHost, " "))
			if subfix != "" {
				//hosts
				if len(u) == 2 {
					(*dnsHosts)[u[1]+subfix] = u[0]
					n++
				}
			} else {
				//dns forward
				if len(u) == 2 && u[1] != "" {
					(*dnsHosts)[u[0]] = u[1]
					last = u[1]
					n++
				} else if len(u) > 0 {
					(*dnsHosts)[u[0]] = last
					n++
				}
			}
		}
		if n > 0 {
			debugf("hosts file %s loaded, %d", hosts, n)
		}
	} else {
		panic(fmt.Errorf("host file not found , %s", hosts))
	}
}
func dnsServer() {
	dns.HandleFunc(".", dnsCallback)
	go func() {
		defer func() {
			if e := recover(); e != nil {
				debugf("callback handler crashed, err : %s\nstack:%s", e, string(debug.Stack()))
			}
		}()
		log.Printf("dns server listen on udp %s", dnsListen)
		dnsListenerServer = &dns.Server{Addr: dnsListen, Net: "udp"}
		err := dnsListenerServer.ListenAndServe()
		if err != nil {
			debugf("dns listen error: %s", err)
		}

	}()
}
func dnsCallback(w dns.ResponseWriter, req *dns.Msg) {
	defer func() {
		if err := recover(); err != nil {
			debugf("dns handler crashed with err : %s \nstack: %s", err, string(debug.Stack()))
		}
	}()
	var (
		key       string
		m         *dns.Msg
		err       error
		data      []byte
		id        uint16
		query     []string
		questions []dns.Question
	)
	if req.MsgHdr.Response == true {
		return
	}
	query = make([]string, len(req.Question))

	for i, q := range req.Question {
		if q.Qtype != dns.TypeAAAA {
			questions = append(questions, q)
		}
		query[i] = fmt.Sprintf("(%s %s %s)", q.Name, dns.ClassToString[q.Qclass], dns.TypeToString[q.Qtype])
	}

	if len(questions) == 0 {
		return
	}

	req.Question = questions
	id = req.Id
	req.Id = 0
	key = toMd5(req.String())
	req.Id = id
	//hosts
	for _, q := range req.Question {
		if q.Qtype == dns.TypeA {
			//log.Printf("q.Name %s %v", q.Name, dnsHosts)
			if v, ok := dnsHosts[q.Name]; ok {
				m := new(dns.Msg)
				m.SetReply(req)
				m.Compress = false
				m.Answer = []dns.RR{}
				rr, _ := dns.NewRR(fmt.Sprintf("%s A %s", q.Name, v))
				m.Answer = append(m.Answer, rr)
				w.WriteMsg(m)
				debugf("id: %5d hosts: HIT %v", id, query)
				return
			}
		}
	}

	//cache
	if reply, ok := cache.Get(key); ok {
		data, _ = reply.([]byte)
	}
	if data != nil && len(data) > 0 {
		m = &dns.Msg{}
		m.Unpack(data)
		m.Id = id
		err = w.WriteMsg(m)
		debugf("id: %5d cache: HIT %v", id, query)
		return
	}

	//resolve
	answer := []byte{}

	//forward resovle
	for _, q := range req.Question {
		if q.Qtype == dns.TypeA {
			dnsAddr := ""
			domain := strings.TrimRight(q.Name, ".")
			if _d, ok := dnsGlobCache.Load(domain); ok {
				dnsAddr = _d.(string)
			} else {
				for k, v := range dnsForward {
					if strings.Index(v, ":") == -1 {
						v = v + ":53"
					}
					g := glob.MustCompile(k, '.')
					//debugf("%s -> %s : %v", k, v, g.Match(domain))
					if g.Match(domain) {
						dnsAddr = v
						break
					}
				}
			}

			if dnsAddr != "" {
				if _, ok := dnsGlobCache.Load(domain); !ok {
					dnsGlobCache.Store(domain, dnsAddr)
				}
				c := new(dns.Client)
				c.Dialer = &net.Dialer{
					Timeout: time.Duration(timeout) * time.Millisecond,
				}
				debugf("use %s relsove %s", dnsAddr, domain)
				m, _, err := c.Exchange(req, dnsAddr)
				if err != nil {
					debugf(err)
				} else {
					answer, _ = m.Pack()
				}
			}
		}
	}
	//nonforward resovle, use parent resovle
	if len(answer) == 0 {
		//use parent resolve
		lbAddr := lb.Select("", loadBalanceOnlyHA)
		debugf("id: %5d resolve: %v %s %s", id, query, lbAddr, dnsServerAddress)
		outconn, err := getOutconn(lbAddr, dnsServerAddress)
		if err != nil {
			debugf("dns query fail, %s", err)
			return
		}
		defer func() {
			outconn.Close()
		}()
		b, _ := req.Pack()
		outconn.Write(append([]byte{0, byte(len(b))}, b...))
		answer, _ = ioutil.ReadAll(outconn)
		if len(answer) < 3 {
			debugf("dns query fail, bad response")
			outconn.Close()
			return
		}
		answer = answer[2:]
	}
	defer func() {
		answer = nil
	}()
	m = &dns.Msg{}
	m.Unpack(answer)
	m.Id = req.Id
	if len(m.Answer) == 0 {
		debugf("dns query fail, no answer")
		return
	}
	d, err := m.Pack()
	if err != nil {
		debugf("dns query fail,%s", err)
		return
	}
	_, err = w.Write(d)
	if err != nil {
		debugf("dns query fail,%s", err)
		return
	}
	ttl := 0
	if len(m.Answer) > 0 {
		if dnsTTL > 0 {
			ttl = dnsTTL
		} else {
			ttl = int(m.Answer[0].Header().Ttl)
			if ttl < 0 {
				ttl = dnsTTL
			}
		}
	}
	cache.Set(key, answer, time.Second*time.Duration(ttl))
	log.Printf("id: %5d cache: CACHED %v TTL %v", id, query, ttl)
}
func toMd5(data string) string {
	m := md5.New()
	m.Write([]byte(data))
	return hex.EncodeToString(m.Sum(nil))
}
func debugf(v ...interface{}) {
	if nolog {
		return
	}
	str := v[0].(string)
	if isDebug {
		log.Printf(str, v[1:]...)
	}
}
func getOutconn(lbAddr, targetAddr string) (outconn net.Conn, err error) {
	if outboundUDP {
		outconn, err = clienttransport.TOUConnectHost(lbAddr, method, getPassword(lbAddr), compress, timeout*1000)
	} else {
		if outboundEncrypt {
			outconn, err = clienttransport.TCPSConnectHost(lbAddr, method, getPassword(lbAddr), compress, timeout*1000)
		} else {
			outconn, err = net.DialTimeout("tcp", lbAddr, time.Duration(timeout)*time.Second)
		}
	}
	if err == nil && targetAddr != "" {
		outconn.Write(utils.BuildPacketData(targetAddr))
	}
	return
}

var passwordMap = map[string]string{}

func getPassword(addr string) string {
	if v, ok := passwordMap[addr]; ok {
		return v
	}
	return password
}
func initLB() {
	configs := lbx.BackendsConfig{}
	for _, addr := range forwardsAddr {
		_addrInfo := strings.Split(addr, "#")
		_addr := ""
		if i := strings.LastIndex(_addrInfo[0], "@"); i >= 0 {
			_addr = _addrInfo[0][i+1:]
			passwordMap[_addr] = _addrInfo[0][:i]
		} else {
			_addr = _addrInfo[0]
		}
		weight := 1
		if len(_addrInfo) == 2 {
			weight, _ = strconv.Atoi(_addrInfo[1])
		}

		configs = append(configs, &lbx.BackendConfig{
			Address:       _addr,
			Weight:        weight,
			ActiveAfter:   loadBalanceActiveAfter,
			InactiveAfter: loadBalanceInactiveAfter,
			Timeout:       time.Duration(loadBalanceTimeout) * time.Millisecond,
			RetryTime:     time.Duration(loadBalanceRetryTime) * time.Millisecond,
		})
	}
	lb = lbx.NewGroup(utils.LBMethod(loadBalanceMethod), configs, nil, log, false)
}
